File: Problems/Algorithms/876. Middle of the Linked List/MiddleLinkedList.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun middleNode(head: ListNode?): ListNode? {
        var fast: ListNode? = head
        var slow: ListNode? = head
        
        while (fast != null && fast?.next != null) {
            fast = fast?.next?.next
            slow = slow?.next
        }
        
        return slow
    }
}

File: Problems/Algorithms/876. Middle of the Linked List/middle_linked_list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut fast = head.clone();
        let mut slow = head.clone();

        while fast.is_some() && fast.as_ref()?.next.is_some() {
            fast = fast.unwrap().next.unwrap().next;
            slow = slow.unwrap().next;
        }

        slow
    }
}

